src/cli/parse-flags.ts

```diff
@@ -54,131 +56,265 @@ export const parseFlags = (args: string[], _sys?: CompilerSystem): ConfigFlags =
+
+  if (readOnlyArrayHasStringMember(BOOLEAN_CLI_FLAGS, normalizedArg)) {
+    // We're setting a boolean!
+    flags[normalizedArg] =
+      typeof value === 'string'
+        ? Boolean(value)
+        : // no value was supplied, default to true
+          true;
+    flags.knownArgs.push(rawArg);
+  } else if (readOnlyArrayHasStringMember(STRING_CLI_FLAGS, normalizedArg)) {
+    // We're setting a string!
+    if (typeof value === 'string') {
+      flags[normalizedArg] = value;
+      flags.knownArgs.push(rawArg);
+      flags.knownArgs.push(value);
+    } else {
+      throwCLIParsingError(rawArg, 'expected a string argument but received nothing');
+    }
+  } else if (readOnlyArrayHasStringMember(STRING_ARRAY_CLI_FLAGS, normalizedArg)) {
+    // We're setting a string, but it's one where the user can pass multiple values,
+    // like `--reporters="default" --reporters="jest-junit"`
+    if (typeof value === 'string') {
+      if (!Array.isArray(flags[normalizedArg])) {
+        flags[normalizedArg] = [];
+      }
 
-  if (value !== undefined && matchingArg !== undefined) {
-    if (CLI_ARG_STRING_REGEX.test(value)) {
-      // if it matches the regex we treat it like a string
-      flags[configCaseName] = value;
+      const targetArray = flags[normalizedArg];
+      // this is irritating, but TS doesn't know that the `!Array.isArray`
+      // check above guarantees we have an array to work with here, and it
+      // doesn't want to narrow the type of `flags[normalizedArg]`, so we need
+      // to grab a reference to that array and then `Array.isArray` that. Bah!
+      if (Array.isArray(targetArray)) {
+        targetArray.push(value);
+        flags.knownArgs.push(rawArg);
+        flags.knownArgs.push(value);
+      }
     } else {
-      // it was a number, great!
-      flags[configCaseName] = Number(value);
+      throwCLIParsingError(rawArg, 'expected a string argument but received nothing');
+    }
+  } else if (readOnlyArrayHasStringMember(NUMBER_CLI_FLAGS, normalizedArg)) {
+    // We're setting a number!
+    if (typeof value === 'string') {
+      const parsed = parseInt(value, 10);
+
+      if (isNaN(parsed)) {
+        throwNumberParsingError(rawArg, value);
+      } else {
+        // parseInt went alright, so here we go
+        // (duck typing?)
+        flags[normalizedArg] = parsed;
+        flags.knownArgs.push(rawArg);
+        flags.knownArgs.push(value);
+      }
+    } else {
+      throwCLIParsingError(rawArg, 'expected a number argument but received nothing');
+    }
+  } else if (readOnlyArrayHasStringMember(STRING_NUMBER_CLI_FLAGS, normalizedArg)) {
+    // We're setting a value which could be either a string _or_ a number
+    if (typeof value === 'string') {
+      if (CLI_ARG_STRING_REGEX.test(value)) {
+        // if it matches the regex we treat it like a string
+        flags[normalizedArg] = value;
+      } else {
+        const parsed = Number(value);
```

Comment:
I want to point out an inconsistency in number handling here. Does it make sense to just use `Number` on the wrong string, or to try to use `parseInt` ? We did have (when I first went in to this module) some idea of supporting scientific notation like `40e4`. That rules out using `parseInt`, which will return `40` for that. Just `Number` on a string, however, does do some other stuff which is maybe undesirable, like `Number("")` is `0`. Anywhoo - I have it one way for a `NUMBER_CLI_FLAG` and another way for this `STRING_NUMBER_CLI_FLAG`, interested in hearing thoughts about which way we should go (another alternative would be carving out a `parseCLINumber` helper function that has some logic to figure out which of the two, `parseInt` or just `Number`, is suitable)

Reply:
I think so (that this makes sense) 👍